src/translator/cache.cpp

```diff
@@ -0,0 +1,168 @@
+#include "translator/cache.h"
+
+#include <cstdlib>
+
+namespace marian {
+namespace bergamot {
+
+#ifndef WASM_COMPATIBLE_SOURCE
+
+ThreadSafeL4Cache::ThreadSafeL4Cache(Ptr<Options> options)
+    : epochManagerConfig_(options->get<size_t>("cache-ebr-queue-size"),
+                          std::chrono::milliseconds(options->get<size_t>("cache-ebr-interval")),
+                          options->get<size_t>("cache-ebr-num-queues")),
+      cacheConfig_(options->get<size_t>("cache-size") * 1024 * 1024,
+                   std::chrono::seconds(options->get<size_t>("cache-time-to-live")),
+                   options->get<bool>("cache-remove-expired")),
+      service_(epochManagerConfig_),
+      context_(service_.GetContext()) {
+  // There is only a single cache we use. However, L4 construction API given by example demands we give it a string
+  // identifier.
+  const std::string cacheIdentifier = "global-cache";
+  size_t numBuckets = options->get<size_t>("cache-buckets");
+  hashTableIndex_ = service_.AddHashTable(L4::HashTableConfig(
+      cacheIdentifier, L4::HashTableConfig::Setting{static_cast<uint32_t>(numBuckets)}, cacheConfig_));
+}
+
+bool ThreadSafeL4Cache::fetch(const marian::Words &words, ProcessedRequestSentence &processedRequestSentence) {
+  auto &hashTable = context_[hashTableIndex_];
+
+  /// We take marian's default hash function, which is templated for hash-type. We treat marian::Word (which are indices
+  /// represented by size_t (unverified) and convert it into a uint64_t for hashing with more bits)
+  ///
+  /// With the use of a 64-bit hash, the probability of collision is as follows:
+  /// 100K sentences: 2.710478508305414e-10
+  ///   1M sentences: 2.7105026889628903e-8
+  ///   1B sentences: 0.026741008873954697
```

Comment:
Those probabilities are wrong/misleading since our cache is not infinite and gets reset often (when the user closes the application).

Reply:
Ended up removing.